src/main/java/net/neoforged/neoforge/codec/IndexedNonNullListCodec.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.codec;
+
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.MapCodec;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import net.minecraft.core.NonNullList;
+
+public record IndexedNonNullListCodec<E>(MapCodec<E> elementCodec, E defaultValue, Predicate<E> skipPredicate, String indexKey) implements Codec<NonNullList<E>> {
```

Comment:
Would you mind explaining the reason for a map codec here? I'm guessing it's because it's being stored as:
```json5
{
    "size": int,
    "items": [
        {
            "index": int,
            "value": obj // Entry here
        }
    ]
}
```

Reply:
Yes, the map codec is required for inlining the other data alongside `<index>`